task/backend/scheduler/treescheduler.go

```diff
@@ -0,0 +1,411 @@
+						continue schedulerLoop
+					}
+					s.process()
+					min = s.scheduled.Min()
+					if min == nil { // grab a new item, because there could be a different item at the top of the queue after processing
+						s.when = time.Time{}
+						s.mu.Unlock()
+						continue schedulerLoop
+					}
+					it = min.(Item)
+					s.when = it.when()
+					until := s.when.Sub(s.time.Now())
+
+					if until > 0 {
+						s.resetTimer(until) // we can reset without a stop because we know it is fired here
+						s.mu.Unlock()
+						continue schedulerLoop
+					}
+					s.mu.Unlock()
+				}
+			}
+		}
+	}()
+	return s, s.sm, nil
+}
+
+func (s *TreeScheduler) Stop() {
+	s.mu.Lock()
+	close(s.done)
+	s.mu.Unlock()
+	s.wg.Wait()
+}
+
+type unsent struct {
+	items []Item
+}
+
+func (u *unsent) append(i Item) {
+	u.items = append(u.items, i)
+}
+
+func (s *TreeScheduler) process() {
+	iter, toReAdd := s.iterator(s.time.Now())
+	s.scheduled.Ascend(iter)
+	for i := range toReAdd.items {
+		s.nextTime[toReAdd.items[i].id] = toReAdd.items[i].ordering
+		s.scheduled.ReplaceOrInsert(toReAdd.items[i])
+	}
+}
+
+func (s *TreeScheduler) resetTimer(whenFromNow time.Duration) {
+	s.when = s.time.Now().Add(whenFromNow)
+	s.timer.Reset(whenFromNow)
+}
+
+func (s *TreeScheduler) iterator(ts time.Time) (btree.ItemIterator, *unsent) {
+	itemsToPlace := &unsent{}
+	return func(i btree.Item) bool {
+		if i == nil {
+			return false
+		}
+		it := i.(Item) // we want it to panic if things other than Items are populating the scheduler, as it is something we can't recover from.
+		if time.Unix(it.next+it.Offset, 0).After(ts) {
+			return false
+		}
+		// distribute to the right worker.
+		{
+			buf := [8]byte{}
+			binary.LittleEndian.PutUint64(buf[:], uint64(it.id))
+			wc := xxhash.Sum64(buf[:]) % uint64(len(s.workchans)) // we just hash so that the number is uniformly distributed
+			select {
+			case s.workchans[wc] <- it:
+				s.scheduled.Delete(it)
+				if err := it.updateNext(); err != nil {
+					// in this error case we can't schedule next, so we have to drop the task
+					s.onErr(context.Background(), it.id, it.Next(), &ErrUnrecoverable{err})
+					return true
+				}
+				itemsToPlace.append(it)
+
+			case <-s.done:
+				return false
+			default:
+				s.scheduled.Delete(it)
+				it.incrementNonce()
+				itemsToPlace.append(it)
+				return true
+			}
+		}
+		return true
+	}, itemsToPlace
+}
+
+// When gives us the next time the scheduler will run a task.
+func (s *TreeScheduler) When() time.Time {
+	s.mu.RLock()
+	w := s.when
+	s.mu.RUnlock()
+	return w
+}
+
+func (s *TreeScheduler) release(taskID ID) {
+	ordering, ok := s.nextTime[taskID]
+	if !ok {
+		return
+	}
+
+	// delete the old task run time
+	s.scheduled.Delete(Item{id: taskID, ordering: ordering})
+	delete(s.nextTime, taskID)
+}
+
+// Release releases a task.
+// Release also cancels the running task.
+// Task deletion would be faster if the tree supported deleting ranges.
+func (s *TreeScheduler) Release(taskID ID) {
```

Comment:
I don't see any tests for `Release`; could you add them and then we're good to go!